Registration table of registered patients

bucephalus opened the pull request:
For this feature, still need to do the following:
- Limit scope of registrations shown
- Show more when "See more" is clicked
- Display notice in table when no registrations have been made (optional?)

I will be working on the last one first.

Also, I'm getting the following error when I visit localhost:3000/registration/registrars:
uninitialized constant RegistrarsController::ChronicDuration
Application Trace:
app/controllers/registrars_controller.rb:72:in `average_time'
app/controllers/registrars_controller.rb:4:in`index'

bucephalus committed:
filled in table-using steps
altered registration model to include patient_name and registrar_name

bucephalus committed:
minor step definition typo fixes

bucephalus committed:
added pending step definitions

bucephalus committed:
removed step definition
planning to change cucumber tests

bucephalus committed:
changed cucumber scenarios
to reflect table functionality from lo-fi mockups

bucephalus committed:
split feature into overall and individual features

bucephalus committed:
edited background for overall case
and some minor edits to  scenario names

bucephalus committed:
for overall, added scenario for format of registration

bucephalus committed:
in overall feature, all registrations made for different patients
  (for clarity in scenarios)
filled in first 4 scenarios

bucephalus committed:
in overall feature, fixed typo in scenario

bucephalus committed:
for overall feature, filled in all scenarios

bucephalus committed:
added pending step definitions based on overall

bucephalus committed:
removed quotation marks from step definitions for seeing list of patients

bucephalus committed:
typo in overall feature

bucephalus committed:
in overall feature
added some step definitions
edited names of steps to not collide with previously defined steps

bucephalus committed:
for registration_steps.rb, filled in step definitions
all definitions for overall case, except for steps dealing with
no registrations
Then I should see patient "patient" with time "time" and status "status"

bucephalus committed:
added feature for veiwing empty table if no registrations

bucephalus committed:
removed old scenarios and "no registrations" scenario from overall feature

bucephalus committed:
copied overall feature into individual feature
changed so all registrars in registrations are Reg A
changed overall registration dashboard to Reg A dashboard

bucephalus committed:
in individual feature, changed some registrations to be by Reg B
changed related scenario to reflect not showing registrations by Reg B

bucephalus committed:
in registration_steps.rb, added definition for no registrations step

bucephalus committed:
added 'autotest-rails' to gemfile

bucephalus committed:
Merge remote-tracking branch 'origin/master' into registration_table_of_registered_patients

Conflicts:
	app/controllers/registrars_controller.rb
	app/models/registration.rb
	app/views/registrars/show.html.haml
	features/registration/patients_table.feature
	features/step_definitions/registration_steps.rb

bucephalus committed:
added @ in front of num_today and average_time in _rightbar.html.haml
page would not render without the change

bucephalus committed:
removed patient_name and registrar_name from registration model

bucephalus committed:
added partial to display registrations table

bucephalus committed:
typos in _regist_table.html.haml

bucephalus committed:
removed a puts from registration_steps.rb

bucephalus committed:
added setup of variable used in partial to registrars_controller.rb

bucephalus committed:
fixed no_match step definition

bucephalus committed:
fixed some typos in registrars_controller.rb

bucephalus committed:
edited registration_steps.rb to match implementation

bucephalus committed:
fixed registrars_controller.rb to not display date before every registration

bucephalus committed:
changed path name in registration_steps.rb
for going to an individual dashboard

bucephalus committed:
minor fixes

bucephalus committed:
in registrars_controller.rb moved setup from index to show
in _regist_table.html.haml fixed minor error

bucephalus committed:
changed features because I derped and thought time was backwards

bucephalus committed:
more similar changes to last

bucephalus committed:
now deals with times as localtime
to prevent strange issues using two timezones

bucephalus committed:
Merge remote-tracking branch 'origin/master' into registration_table_of_registered_patients

Conflicts:
	features/registration/patients_table.feature

bucephalus committed:
removed debugging puts from step definition

bucephalus committed:
moved registration table partial into registrars folder
made associated changes to code path used in show
removed old table code from show

bucephalus committed:
added back partial for regist table

bucephalus committed:
edited registrars_controller so that
it only filters by registrar id if one is given

bucephalus committed:
added step definition to check patient listed at correct time/status

bucephalus committed:
fixed typos in new step in registration_steps.rb

bucephalus committed:
set up @registrations_and_divs in index of registrars_controller.rb

bucephalus committed:
removed autotest-rails from gemfile

bucephalus committed:
reran bundle install

bucephalus committed:
added code to implement "no registrations" line in table
added to registrars_controller.rb and a conditional to the partial

breaking: average_time may not know how to deal with no registrations
and returns nil?

bucephalus committed:
in average_time, set total_time to 0 if no registrations

bucephalus committed:
conditionally return 0 from average_time if there are no registrations
no longer set total_time to 0

bucephalus committed:
typo: originally set @registrations_and_divs to [nil, message]
corrected: now add [nil, message] to @registrations_and_divs

bucephalus commented:
added stuff for no registrations

bucephalus commented:
The ChronicDuration error no longer happens.  I don't know why.

bucephalus committed:
in registrars_controller.rb, for show and index,
when setting up @registrations_and_divs,
reverse order using DB instead of using Array#reverse

bucephalus committed:
DRYed out creation of @registrations_and_divs in registrars_controller.rb

seshness commented on app/controllers/registrars_controller.rb:
```diff
@@ -2,6 +2,33 @@ class RegistrarsController < ApplicationController
+    #    and the second element indicates whether or not to place a divider
+    #        before the next registration in the table
+    #        using false for no divider, and the divider text otherwise
+    @registrations = Registration.all(:order => "time_end")
```
Move the `.reverse` below from the Ruby/Rails layer into the DB layer.  See http://guides.rubyonrails.org/active_record_querying.html#ordering

```
@registrations = Registration.all(:order => "time_end DESC")
```

seshness commented on app/controllers/registrars_controller.rb:
```diff
@@ -2,6 +2,33 @@ class RegistrarsController < ApplicationController
+    @registrations = Registration.all(:order => "time_end")
+    @registrations_and_divs = []
+    last_reg = Chronic.parse('today').localtime.strftime('%^B %e, %Y')
+    @registrations.reverse.each do |reg|
```
No `reverse` here then

seshness commented on app/controllers/registrars_controller.rb:
```diff
@@ -2,6 +2,33 @@ class RegistrarsController < ApplicationController
+    #    two-element-lists where the first element is a registration
+    #    and the second element indicates whether or not to place a divider
+    #        before the next registration in the table
+    #        using false for no divider, and the divider text otherwise
```
What if dividers were simply another element in the list? It seems simpler to have a list structured 

```
[a, b, c, DIVIDER('YESTERDAY'), d, e, f]
```

instead of the current

```
[[a, false], [b, false], [c, 'YESTERDAY']], [d false], [e, false], [f, false]]
```

bucephalus replied:
I considered that implementation.  In the view, I'd need to be able to differentiate dividers from registrations, and I'm not sure how.

Also, could you clarify why you represented dividers as DIVIDER('YESTERDAY')?
And, I think the current looks more like:
[[a, false], [b, false], [c, "YESTERDAY"], [d false], [e, false], [f, false]]

bucephalus replied:
I also think I chose this because it was easier to come up with variable names that made sense :P  Not a big deal.

danwang replied:
^ Read above. With my format, you can `each` through the array and process each segment individually.

danwang commented on app/controllers/registrars_controller.rb:
```diff
@@ -2,6 +2,33 @@ class RegistrarsController < ApplicationController
   def index
     @num_today = Registration.for_day(Time.now).count
     @average_time = average_time(Registration.all)
+
```
Instead of having a two-element list, how about a list for every "segment"?

For example, if we wanted to have something like

```
[DATE 1]
PATIENT 1
PATIENT 2
PATIENT 3
[DATE 2]
PATIENT 4
PATIENT 5
```

We could have something like

``` ruby
[ {
  :date => DATE1,
  :patients => [
    PATIENT 1,
    PATIENT 2,
    PATIENT 3,
  ] },
 ...
]
```

bucephalus committed:
changed representation of @registrations_and_divs
in registars_controller.rb

danwang commented on app/controllers/registrars_controller.rb:
```diff
@@ -9,11 +36,43 @@ def show
     @num_today = registrar.registrations_for_day(Time.now).count
     @average_time = average_time(registrar.registrations)
+
+    # set up @registrations_and_divs for registrations table in view
```
DRY this segment out

seshness commented on app/controllers/registrars_controller.rb:
```diff
@@ -9,11 +36,43 @@ def show
+    if @registrations_and_divs == []
+      @registrations_and_divs << [nil, "there are no registrations to show"]
+    end
+
```
Duplicated code smell? Please factor out common logic frmo `index` and `show`

bucephalus committed:
typo fix (used = instead of => in hashtable)

danwang commented on app/controllers/registrars_controller.rb:
```diff
@@ -9,11 +36,43 @@ def show
   protected
   def average_time(registrations)
     total_time = registrations.map {|r| r.elapsed_time}.reduce(:+)
-    return ChronicDuration.output((total_time/registrations.length).to_i, :format => :chrono)
```
Use `reduce(0, :+)` instead for initial value

bucephalus replied:
the actual issue is that registrations.length will return 0 if there are no registrations, and thus divide by zero

danwang commented on app/views/registrars/_regist_table.html.haml:
```diff
@@ -0,0 +1,26 @@
+          %td.divider{:colspan => 3}
+            %div
+              %span{:title => div}= div
+      -if registration
```
If we restructure the registration data, this won't be necessary

seshness commented on features/registration/patients_table_individual.feature:
```diff
@@ -0,0 +1,103 @@
+    But I should not see the following patients: Patient T,Patient U,Patient V
+
+Scenario: see more, shows more registrations
+    When I click SEE MORE
```
Let's stick to the more conventional

```
When I click "See More"
```

bucephalus replied:
should I put this in a new step file, because it's so general?  or can I just put it in registration_steps.rb?

seshness commented on features/registration/patients_table_individual.feature:
```diff
@@ -0,0 +1,103 @@
+    And I should see the following patients: Patient J,Patient P,Patient Q
+    And I should see the following patients: Patient R,Patient S,Patient T
+    And I should see the following patients: Patient U,Patient V
+    But I should not see the following patients: Patient K,Patient L, Patient M, Patient N, Patient O
```
inconsistent spaces after commas

danwang commented on app/views/registrars/_rightbar.html.haml:
```diff
@@ -1,12 +1,12 @@
   %div.bignum#num-today
-    = num_today
-  = "#{'patient'.pluralize(num_today)} registered today"
+    = @num_today
```
IMO, it's better to use the variable `num_today` than relying on the parent view to have a `@num_today` available.

bucephalus committed:
changed view to work with new representation
fixed some logic in get_regs_by_date

bucephalus committed:
forgot to change some "div"s in view to regs_by_date[:date]

danwang commented on app/views/registrars/_regist_table.html.haml:
```diff
@@ -0,0 +1,26 @@
+        %div
+          -# please change the title attribute of the next line
+          %span
+            %a{:href => '#'} See more
```
Let's use the awesome Rails functions like `link_to` instead of dry `a` elements

danwang commented on app/views/registrars/_regist_table.html.haml:
```diff
@@ -0,0 +1,26 @@
+      %th Patient
+      %th Status
+  %tbody
+    - @registrations_and_divs.each do |registration, div|
```
Same thing here -- pass in `registrations_and_divs` as a local variable instead of using the instance variable

danwang commented on app/views/registrars/show.html.haml:
```diff
@@ -5,53 +5,5 @@
-        %span
-          %a{:href => '#'} See more
+
+= render :partial => 'registrars/regist_table'
```
`:locals => .....`

danwang commented on features/step_definitions/registration_steps.rb:
```diff
@@ -70,3 +70,55 @@
+# table of registered patients
+
+Then /^I should see patient "(.*?)" with time "(.*?)" and status "(.*?)"$/ do |patient_name, time, status|
+  html_string = "<tr>\n<td> #{time}</td>\n<td>#{patient_name}</td>\n<td>#{status}</td>\n"
```
This seems very bad. It's very view-dependent. What if we decide to add a class name to a `tr` or `td`? Then this has to change. Use something that looks at content instead.

danwang commented on features/step_definitions/registration_steps.rb:
```diff
@@ -70,3 +70,55 @@
+Then /^I should see the following patients: (.*)$/ do |patients_list|
+  patients_list = patients_list.split(/,/)
+  patients_list.each do |patient_name|
+    assert_match(/#{patient_name}/m, page.body)
```
Again, instead of using regex match, do something like assign class/ID to the relevant able elements and search the content for the patient name

danwang commented on features/step_definitions/registration_steps.rb:
```diff
@@ -70,3 +70,55 @@
+Then /^I should not see the following patients: (.*)$/ do |patients_list|
+  patients_list = patients_list.split(/,/)
+  patients_list.each do |patient_name|
+    assert_no_match(/#{patient_name}/m, page.body)
```
Same as above

danwang commented on features/step_definitions/registration_steps.rb:
```diff
@@ -70,3 +70,55 @@
+  end
+end
+
+When /^I click SEE MORE$/ do
```
No need for this definition. Why not make it more general with a regex? If we followed this style, we'd have one step definition for every possible button

danwang commented on features/step_definitions/registration_steps.rb:
```diff
@@ -70,3 +70,55 @@
+end
+
+Then /^I should see a no registrations notification$/ do
+  assert_match("there are no registrations to show", page.body)
```
Attach an id/class to where the notification should show up

danwang commented on features/registration/patients_table_individual.feature:
```diff
@@ -0,0 +1,103 @@
+      | 7 days ago 10:00 am | 7 days ago 11:00 am | returning      | Patient S    | Reg A          |
+      | 7 days ago 11:00 am | 7 days ago 1:00 pm  | new            | Patient T    | Reg A          |
+      | 7 days ago 1:00 pm  | 7 days ago 2:00 pm  | returning      | Patient U    | Reg A          |
+      | 7 days ago 3:00 pm  | 7 days ago 5:00 pm  | returning      | Patient V    | Reg A          |
```
DRY all of this out -- lots of duplicate between individual and non-individual

danwang commented on features/registration/patients_table_overall.feature:
```diff
@@ -0,0 +1,114 @@
+    And I should see patient "Patient B" with time "2:00 pm" and status "new"
+
+Scenario: by default, show last 10 patients registered
+    Then I should see the following patients: Patient A,Patient B,Patient C
```
Can we do something like set the variable `@all_patients` in `Given the following patients exist` and then using that in steps like

`Then I should see all of the patients`

and

`Then I should see all of the patients except ....`

Low-priority.

bucephalus committed:
changed all 11:00 am to 10:00 am in idividual feature
because of strange error with Chronic::parse('11:00 am') not working

bucephalus committed:
set up curr_regs to have empty :regs for today as well
in registrars_controller.rb

bucephalus committed:
changed all '11:00 pm' in individual to '10:30 pm'
because of chronic

bucephalus committed:
in overall feature, replaced 11:00 am with 10:30 am
and replaced 11:00 pm with 10:30 pm

bucephalus committed:
use more general: When I click the link "(.*?)"
instead of: When I click SEE MORE
changed step definitions, overall & individual features

bucephalus committed:
fixed inconsistent commas in individual feature
for But I should not see...

bucephalus committed:
in _rightbar partial view, changed num_today and average_time back to local vars

bucephalus committed:
in regist_table partial, use link_to instead of a href
link to nil instead of #

bucephalus committed:
regist_table partial now uses local variable for registrations_and_divs
passed in by show view

erictzeng commented:
This pull request is kinda sorta working, and it's way too big now. Gonna merge.
